compiler/plugins/target/ROCM/builtins/ukernel/argmax_ukernel.c

```diff
@@ -5,14 +5,50 @@
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
 #include <float.h>
-#include <hip/hip_fp16.h>
-#include <hip/hip_runtime.h>
+#include <stdint.h>
+#include <string.h>
 
-extern "C" __device__ __attribute__((const)) half __ockl_wfred_max_f16(half);
-extern "C" __device__
-    __attribute__((const)) int64_t __ockl_wfred_min_i64(int64_t);
-extern "C" __device__
-    __attribute__((const)) int32_t __ockl_wfred_min_i32(int32_t);
+/*
+Bits copied from HIP headers.
+*/
+
+__attribute__((const)) _Float16 __ockl_wfred_max_f16(_Float16);
+__attribute__((const)) int64_t __ockl_wfred_min_i64(int64_t);
+__attribute__((const)) int32_t __ockl_wfred_min_i32(int32_t);
+__attribute__((const)) float __ocml_fmax_f32(float, float);
+__attribute__((const)) float __ocml_fmax_f16(_Float16, _Float16);
+
+static inline unsigned int __lane_id() {
+  return __builtin_amdgcn_mbcnt_hi(-1, __builtin_amdgcn_mbcnt_lo(-1, 0));
+}
+
+static inline int __shfl_xor_i(int var, int lane_mask) {
+  const int width = __builtin_amdgcn_wavefrontsize();
+  int self = __lane_id();
+  int index = self ^ lane_mask;
+  index = index >= ((self + width) & ~(width - 1)) ? self : index;
+  return __builtin_amdgcn_ds_bpermute(index << 2, var);
+}
+
+static inline float __shfl_xor_f(float var, int lane_mask) {
+  union {
+    int i;
+    unsigned u;
+    float f;
+  } tmp;
+  tmp.f = var;
+  tmp.i = __shfl_xor_f(tmp.i, lane_mask);
+  return tmp.f;
```

Comment:
Cool stuff! just curious, where did you get this `__shfl_xor_f` implementation from (can't seem to spot it in LLVM or my hip includes)? Also do they not have same thing for `shfl_xor_i`?

Reply:
It's in hipamd/include/hip/amd_detail/amd_warp_functions.h from https://github.com/ROCm/clr.  It's named just `__shfl_xor` here for various element types because HIP allows function overloading; the suffixes here are my addition.